src/BaseControllerV2.test.ts

```diff
@@ -504,4 +564,139 @@ describe('getPersistentState', () => {
+      type: `VisitorController:stateChange`;
+      payload: [VisitorControllerState, Patch[]];
+    };
+    const visitorControllerStateMetadata = {
+      visitors: {
+        persist: true,
+        anonymous: true,
+      },
+    };
+    class VisitorController extends BaseController<'VisitorController', VisitorControllerState> {
+      constructor(
+        messagingSystem: RestrictedControllerMessenger<
+          'VisitorController',
+          VisitorControllerAction | VisitorOverflowControllerAction,
+          VisitorControllerEvent | VisitorOverflowControllerEvent,
+          never,
+          never
+        >,
+      ) {
+        super(messagingSystem, 'VisitorController', { visitors: [] }, visitorControllerStateMetadata);
+        messagingSystem.registerActionHandler('VisitorController:clear', this.clear);
+      }
+
+      clear = () => {
+        this.update(() => {
+          return { visitors: [] };
+        });
+      };
+
+      addVisitor(visitor: string) {
+        this.update(({ visitors }) => {
+          return { visitors: [...visitors, visitor] };
+        });
+      }
+
+      destroy() {
+        super.destroy();
+      }
+    }
+
+    type VisitorOverflowControllerState = {
+      maxVisitors: number;
+    };
+    type VisitorOverflowControllerAction = {
+      type: `VisitorOverflowController:updateMax`;
+      handler: (max: number) => void;
+    };
+    type VisitorOverflowControllerEvent = {
+      type: `VisitorOverflowController:stateChange`;
+      payload: [VisitorOverflowControllerState, Patch[]];
+    };
+
+    const visitorOverflowControllerMetadata = {
+      maxVisitors: {
+        persist: false,
+        anonymous: true,
+      },
+    };
+
+    class VisitorOverflowController extends BaseController<
+      'VisitorOverflowController',
+      VisitorOverflowControllerState
+    > {
+      constructor(
+        messagingSystem: RestrictedControllerMessenger<
+          'VisitorOverflowController',
+          VisitorControllerAction | VisitorOverflowControllerAction,
+          VisitorControllerEvent | VisitorOverflowControllerEvent,
+          'VisitorController:clear',
+          'VisitorController:stateChange'
+        >,
+      ) {
+        super(messagingSystem, 'VisitorOverflowController', { maxVisitors: 5 }, visitorOverflowControllerMetadata);
+        messagingSystem.registerActionHandler('VisitorOverflowController:updateMax', this.updateMax);
+        messagingSystem.subscribe('VisitorController:stateChange', this.onVisit);
+      }
+
+      onVisit = ({ visitors }: VisitorControllerState) => {
+        if (visitors.length > this.state.maxVisitors) {
+          this.messagingSystem.call('VisitorController:clear');
+        }
+      };
+
+      updateMax = (max: number) => {
+        this.update(() => {
+          return { maxVisitors: max };
+        });
+      };
+
+      destroy() {
+        super.destroy();
+      }
+    }
+
+    it('should allow messaging between controllers', () => {
```

Comment:
This is the most complete example so far. It demonstrates calling an action and subscribing to an event in another controller.

Reply:
I think I love it. 😍